src/basilisp/core.lpy

```diff
@@ -6589,15 +6589,21 @@
   (let [ctor-name (with-meta
                     (symbol (str "->" (name type-name)))
                     (meta type-name))
+        ctors     [type-name ctor-name]
 
         {:keys [interfaces methods]} (collect-methods method-impls)]
     `(do
+       (declare ~@ctors)
        (deftype* ~type-name ~fields
          :implements [~@interfaces
                       ~'basilisp.lang.interfaces/IType
                       ~'python/object]
          ~@methods)
        (def ~ctor-name ~type-name)
+
+       ~@(for [ctor ctors]
+           `(alter-meta! (var ~ctor) dissoc :redef))
```

Comment:
I'm not sure that the `alter-meta!` step is required. `def` and `deftype*` overwrite the metadata attached to the Vars.